### PR TITLE
check for newer version

### LINKED ISSUE #57: Check for Ajour update
As a user, I would like to know when a new version of Ajour is available.

Suggested solution:
When the app is launched, check if its version matches the one tagged "Latest release" on github. If it does not match: display a "New version available!" clickable button/text, left of the version in the UI.

If the user clicks the button, open https://github.com/casperstorm/ajour/releases/latest in the default browser.

Optional: Add option to disable this check in settings.

### PR DESCRIPTION
**Pull request by tarkah**
resolves #57 

I couldn't find a hyperlink widget in `Iced` for the user to click to take them to the latest release page. For now it just replaces the version with update text styled as an error.

### CODE REVIEW

**Comment by tarkah** [comment]
image

File: src/main.rs [hunk 1]
```diff
@@ -17,6 +17,8 @@ mod wowinterface_api;
 
 use crate::error::ClientError;
 
+pub const VERSION: &str = "0.2.1"; //env!("CARGO_PKG_VERSION");
```
Comment (tarkah) [inline src/main.rs:20]: For testing. Remove before merging.

**Comment by tarkah** [comment]
Latest commit adds a button to open the latest release in browser. For some reason this is panicking, can you test on your end?

File: src/gui/element.rs [hunk 2]
```diff
@@ -410,13 +414,20 @@ pub fn menu_container<'a>(
...
 
-    let version_text = Text::new(env!("CARGO_PKG_VERSION"))
-        .size(DEFAULT_FONT_SIZE)
-        .horizontal_alignment(HorizontalAlignment::Right);
-    let version_container = Container::new(version_text)
-        .center_y()
-        .padding(5)
-        .style(style::SecondaryTextContainer);
+    let version_text = Text::new(if let Some(new_version) = needs_update {
+        format!("New version available {} -> {}", VERSION, new_version)
```
Comment (casperstorm) [inline src/gui/element.rs:418]: Let's write "New Ajour version available" to make it explict that it's a new application update which is available.

**Comment by casperstorm** [comment]
This is GREAT @tarkah. Thank you 🙏🏻.
We should get this PR merged ASAP. A few notes:

> Latest commit adds a button to open the latest release in browser. For some reason this is panicking, can you test on your end?

This has been tested, and does not panic for me on macOS, but does panic on Windows.
I am away from my Windows machine until tomorrow. If you have not identified the bug by then i'll debug it on my machine.

We should also just fix the small conflicts which has arised.

File: src/main.rs (hunk 1 above)
Comment (casperstorm) [reply src/main.rs:20]: This was great for testing. Made it easier.
You can remove this now in a commit.

**Comment by tarkah** [comment]
Gahh, gotta love winapi. I think I found a different crate to handle opening a URL in the default browser, so I'll try that out sometime this weekend. And I'll update the description as you commented.

**Comment by tarkah** [comment]
I was going to try this https://github.com/amodm/webbrowser-rs. But I don't have high hopes because it's using the same ShellExecuteW call under the hood. Must be a problem calling that inside the winit event loop or something. Will have to see if any issues are open in winit that are similar.

**Comment by casperstorm** [comment]
You can try to do like I did with the file dialogue:


https://github.com/casperstorm/ajour/blob/df60035a87e5cba099b382b4c400416c1e4b70bc/src/gui/update.rs#L95

I am away from my main setup until tomorrow, then I'll be able to help you with this PR if we still need a solutions.

**Comment by casperstorm** [comment]
@tarkah do you want me to help you look at this?

**Comment by tarkah** [comment]
Sorry, been helping my brother move all weekend. I'll look into this tomorrow. If you wanna take a stab, feel free to push to my branch.

**Comment by casperstorm** [comment]
> Sorry, been helping my brother move all weekend. I'll look into this tomorrow. If you wanna take a stab, feel free to push to my branch.

What a good brother you are 😀. No worries, take your time. I'll begin some work now on optimizations. With this and that we are in a great spot.

**Comment by tarkah** [comment]
Check out my latest commit, I was able to get it working without panic on windows. Calling that link inside the async runtime instead of the update loop seems to have fixed it. I had to create a message that takes no input, let me know if you want to rename it or anything.

File: src/gui/mod.rs [hunk 3]
```diff
@@ -109,10 +118,12 @@ impl Application for Ajour {
...
     fn new(_flags: ()) -> (Self, Command<Message>) {
-        (
-            Ajour::default(),
+        let init_commands = vec![
             Command::perform(load_config(), Message::Parse),
-        )
+            Command::perform(needs_update(), Message::NeedsUpdate),
+        ];
+
+        (Ajour::default(), Command::batch(init_commands))
```
Comment (casperstorm) [inline src/gui/mod.rs:126]: Great idea to send a batch command initially.

File: src/gui/mod.rs [hunk 4]
```diff
@@ -57,6 +60,8 @@ pub enum Message {
     Error(ClientError),
     UpdateDirectory(Option<PathBuf>),
     FlavorSelected(Flavor),
+    NeedsUpdate(Result<Option<String>>),
+    None(()),
```
Comment (casperstorm) [inline src/gui/mod.rs:64]: I think it's fine to have it. We might need to use it in the future.

File: src/gui/mod.rs [hunk 5]
```diff
@@ -74,6 +79,8 @@ pub struct Ajour {
     ignored_addons: Vec<(Addon, button::State)>,
     is_showing_settings: bool,
     shared_client: Arc<HttpClient>,
+    needs_update: Option<String>,
```
Comment (casperstorm) [inline src/gui/mod.rs:82]: If this was not a typed language it would assume it was a `bool`, alone because of the wording.

Comment (tarkah) [reply src/gui/mod.rs:82]: True... maybe `newer_version`?

**Comment by casperstorm** [comment]
Great addition, @tarkah.
And again, thank you so much for joining this project, it's awesome.
I don't think the hack is that nasty, but maybe we can optimize those things at somepoint now that we are closing in on all features.

Great PR 🎉

**Review by casperstorm — APPROVED** [review]

**Pull request merged**